### ISSUE TITLE
Dependency Handling for a distributed project setup

### ISSUE THREAD

**Issue by shivenduverma-sap** [open-component-model/ocm#448]
Each project has dependencies on other modules which are specified in go.mod files. These dependencies can be categorized into 2 categories and each category has its caveats.
1. Common Dependency Handling: There are a few dependencies which are present in all projects, and we refer to specific versions of a dependency. This created 2 problems:
  - Using a specific old version: Anyone trying to develop has to revert the dependencies to older version. Due to the indirect dependency, it might require the correction in ~ 9 projects and all of the projects might not even be on the same version. Eg: go-container-registry versions
      - https://github.com/open-component-model/replication-controller/blob/main/go.mod#L7
      - https://github.com/open-component-model/ocm-e2e-framework/blob/main/go.mod#L140
      - https://github.com/open-component-model/unpacker-controller/blob/main/go.mod#L89
      - https://github.com/open-component-model/ocm-controllers-sdk/blob/main/go.mod#L10
      - https://github.com/open-component-model/mpas-product-controller/blob/main/go.mod#L9
      - https://github.com/open-component-model/ocm-controller/blob/main/go.mod#L7
      - https://github.com/open-component-model/remote-controller/blob/main/go.mod#L88
      - https://github.com/open-component-model/git-controller/blob/main/go.mod#L7
     
  - Using different dependency versions for different project. Ideally, they all should be on the same version  to get the same functionality and to avoid detecting failing e2e, etc. tests. Eg: ocm-controller versions
      - https://github.com/open-component-model/mpas-product-controller/blob/main/go.mod#L29
      - https://github.com/open-component-model/remote-controller/blob/main/go.mod#L11
      - https://github.com/open-component-model/replication-controller/blob/main/go.mod#L19
      - https://github.com/open-component-model/git-controller/blob/main/go.mod#L22
      - https://github.com/open-component-model/ocm-e2e-framework/blob/main/go.mod#L283
      - https://github.com/open-component-model/unpacker-controller/blob/main/go.mod#L15 

  - Solutions:
    - Have a github actions gather common dependencies from all projects and prevent merging if common dependencies are not on the same/latest version. The versions which should be used should be compiled by the action and stored in a file so that someone who start developing knows the new version they can use locally while developing rather than waiting for the PR to fail and correct before merge
    - Open to other solutions

2. ocm cli dependency: The ocm-cli is referred by all projects and apart from the above problem of using different versions, when ever there are changes(refactoring/features) in the CLI, the failures in other projects should be identified and fixed/reported. The failure should not be a landmine for someone starting a new ticket and the failure-fix should be part of the already identified work done on the cli
  - Refactoring: If changes are compatible or not, the only solution to detect breaking changes is to download all projects locally, dependencies in the projects need to refer to local ocm, and then run the tests. It is a task in itself
  - New Feature: There should be a version description/commit indicating the last stable ocm-cli compatible with all projects or use the approach of problem 1
 Proposed Solution:
  - Go recommends modularizing projects and having a single go.mod file. But since each project has to be able to run in bundles {mpas, mpas-prod, mpas-proj, git-cont, replication-cont, ocm-contr, ocm-cli}, {ocm-cli, ocm-controller, ocm-e2e-f}, {ocm-cli} modularization will not work.
  - Following a non-go approach, where we only commit go.mod file and NOT the go.sum(.gitignore) file. Then improve the central CI pipeline to make it a true CI:
    - Consume a linux Golden image from the multi-cloud team, create a docker-file and install all required dependencies to be able to run go mod tidy and go build
    - Use sap-piper-library or any container-execution functionality to run go commands(go mod tidy, go build) in the container
    - The go.sum is generated for the project in the pipeline and commited to git
    - When there is any change in the cli, the go.mod for each project is edited to use the latest ocm-cli, and the pipeline will download the dependencies and add the go.sum and that will trigger the e2e,etc. tests for each project in the corressponding PRs and we would be able to see the failures at a central place. This is true CI
    - If fix is small then fix it, if complicated, create follow up ticket with the logs from the pipeline

**Comment by Skarlso** [comment]
👋 

Let me try and clarify how dependencies in Go projects work. I highly recommend reading the [go modules reference](https://go.dev/ref/mod) guide in order to fully understand how dependencies are calculated and resolved.

To that end, let me try to clear up some confusion.

> Common Dependency Handling: There are a few dependencies which are present in all projects, and we refer to specific versions of a dependency. This created 2 problems:

These are `replace` statements that are necessary because of OCM, not because of the controllers. OCM deals with a couple of dependencies that require a specific version. These are intrinsic dependencies, and we pin these versions to avoid compiler errors. No one has to revert to anything. These versions already exist on your machine. And you can't use a different one even if you wanted because the project simply wouldn't compile.

This issue is not unique to us. This is a Go based "problem". Even big projects like Kubernetes have `replace` statements.

>Using different dependency versions for different projects. Ideally, they all should be on the same version to get the same functionality and to avoid detecting failing e2e, etc. tests. Eg: ocm-controller versions

Unpacker and remote controller are old. We simply didn't have the time to update them yet.

> Have a github actions gather common dependencies from all projects and prevent merging if common dependencies are not on the same/latest version. The versions which should be used should be compiled by the action and stored in a file so that someone who start developing knows the new version they can use locally while developing rather than waiting for the PR to fail and correct before merge

I think that's not necessary. Especially since the mod files will force the versions, they need over the components they dependent upon. If there is a different version requested, they will simply not compile.

> ocm cli dependency: The ocm-cli is referred by all projects and apart from the above problem of using different versions, when ever there are changes(refactoring/features) in the CLI, the failures in other projects should be identified and fixed/reported. The failure should not be a landmine for someone starting a new ticket and the failure-fix should be part of the already identified work done on the cli

This problem exists in massive projects that depend on a single source. Just look at CAPA, CAPZ, CAPI projects that depend on Kubernetes. They each have cycles where they update their respective Kubernetes version. And get this. Sometimes they support different versions. ;)

We are a small team. It takes me a full day to update all controllers to the latest OCM version because, usually, it introduces some change that breaks existing behaviour. If it doesn't, it's still at least half a day of work updating the dependencies, approving the PRs, and releasing a new version. This is nothing new, though. And it would be nice to automate it. But the reality is that it's simply not feasible considering the usual changes OCM introduces.

Maybe once we separate the CLI from the Library, it will be an easier task to do so.

>Following a non-go approach, where we only commit go.mod file and NOT the go.sum(.gitignore) file. Then improve the central CI pipeline to make it a true CI:

This will simply not work and would prove a huge security risk for local development. Please read the attached Module Reference guide to understand the need for the sum file. On top of that, we would like to avoid non-go approaches, please. :) The generated sum file could provide different SHAs from what the developer is using. This, in-turn, could provide functionality problems, or conflicts during commits which would lead to a frustrating workflow. Further, `go mod tidy` also modifies the mod file. This would mean if the developer isn't running that correctly, that's a further problem in ci when its trying to download and update the sum file.

But the two main concerns are this:
>The go.sum is generated for the project in the pipeline and commited to git

How would this work if it's ignored?

And second, if there is no sum file, or a generated sum file in the repository, that would cause problems with contributors checking out the code. If there is no sum file, they could generate a completely different sum file which is a security risk. Second if, there IS a sum file but it's different from what is generated that's a security issue because the generated sum file could be missing a CVE fix.

**Comment by shivenduverma-sap** [comment]
I think we are not on the same page here on both the issues. Let me clarify

> These are replace statements that are necessary because of OCM, not because of the controllers. OCM deals with a couple of dependencies that require a specific version. These are intrinsic dependencies, and we pin these versions to avoid compiler errors. No one has to revert to anything. These versions already exist on your machine. And you can't use a different one even if you wanted because the project simply wouldn't compile.
> This issue is not unique to us. This is a Go based "problem". Even big projects like Kubernetes have replace statements.
> Unpacker and remote controller are old. We simply didn't have the time to update them yet.

Firstly, everything under the OCM org has to be maintained. Which project introduces what problem is not a solution, how to handle it is the solution. 
The point is that a script will pickup all direct dependencies under require() from all go.mods. And then it will compare if all projects are using the same version of a mentioned dependency. That is it, it makes sure that the versions are the same. If you have a replace overriding it, it doesnt matter, because once the replace is removed, it refers to the same dependency. It is required because some dependency might slip through in the release process, like it did for unpacker&remote controllers and thus it is not updated. Which means that tests might not uncover failures.

This is the only purpose, to solve the problem of handling ~9 go.mod files. In an open source project with external contributors, what dependency is being used where is hard to track. That is the only point here. We are not building kubernetes and the complexity here is much smaller and should be managed. IF you still dont want to do it for all dependencies, just do it for the controllers/ocm-cli versions in go.mod.


> We are a small team. It takes me a full day to update all controllers to the latest OCM version because, usually, it introduces some change that breaks existing behaviour. If it doesn't, it's still at least half a day of work updating the dependencies, approving the PRs, and releasing a new version. This is nothing new, though. And it would be nice to automate it. But the reality is that it's simply not feasible considering the usual changes OCM introduces.
> Maybe once we separate the CLI from the Library, it will be an easier task to do so.
> This will simply not work and would prove a huge security risk for local development. Please read the attached Module Reference guide to understand the need for the sum file. On top of that, we would like to avoid non-go approaches, please. :) The generated sum file could provide different SHAs from what the developer is using. This, in-turn, could provide functionality problems, or conflicts during commits which would lead to a frustrating workflow. Further, go mod tidy also modifies the mod file. This would mean if the developer isn't running that correctly, that's a further problem in ci when its trying to download and update the sum file.
> But the two main concerns are this:
> The go.sum is generated for the project in the pipeline and commited to git
> How would this work if it's ignored?
>And second, if there is no sum file, or a generated sum file in the repository, that would cause problems with contributors checking out the code. If there is no sum file, they could generate a completely different sum file which is a security risk. Second if, there IS a sum file but it's different from what is generated that's a security issue because the generated sum file could be missing a CVE fix.


Again the proposal is not understood correctly. In your projects, you will have a go.sum and will commit and push it there. Nothing changes there.

We will have a separate linux image to compile projects and run tests, where we will not download the go.sum file from the repositories but create a new temporary one.
Why are we doing this? Any changes to the CLI can lead to failures in business cases which needs to be detected by the mpas e2e tests and ocm e2e tests. The only way this can be recognized is when a developer who made the changes to the CLI, downloads all projects offline, updates go.mod, generates new go.sum and runs the tests locally or creates a PR with the tests will run in the pipeline. This means that they have to edit 8 go.mod files for mpas e2e tests. Something which you can not impose on external contributors.

Instead, an automation will do this process and notify of a failure due to the changes. Why do we need an automation and why cant we rely on a developer to detect the changes when they update the cli in controllers to the latest version? Because nobody wants landmines and wasting a day before starting a new ticket.
If you look at how open source development works at Apache, the tests are run on the code before the Apache collaborator even reviews the code that is written by an external contributor for a ticket. We need this process.

**Comment by Skarlso** [comment]
>Firstly, everything under the OCM org has to be maintained. Which project introduces what problem is not a solution, how to handle it is the solution.

Feel free to update them at any time. You are part of this organization.

>And then it will compare if all projects are using the same version of a mentioned dependency

This is not always required.

I'm not sure you understand what these replace statements are doing. They aren't something you can require. They are brought in by transitive dependencies. We replace them BECAUSE we can't require them at a specific version. This is how go modules work.

>In an open source project with external contributors, what dependency is being used where is hard to track.

That is literally what the go.mod file is for.

>Again the proposal is not understood correctly.

This sentence `Following a non-go approach, where we only commit go.mod file and NOT the go.sum(.gitignore) file.` is pretty hard to misinterpret.

>We will have a separate linux image to compile projects and run tests, where we will not download the go.sum file from the repositories but create a new temporary one.

Please search and understand how supply chain attacks work. This is not possible and should be avoided at ALL COST. 


>Why are we doing this? Any changes to the CLI can lead to failures in business cases which needs to be detected by the mpas e2e tests and ocm e2e tests. The only way this can be recognized is when a developer who made the changes to the CLI, downloads all projects offline, updates go.mod, generates new go.sum and runs the tests locally or creates a PR with the tests will run in the pipeline. This means that they have to edit 8 go.mod files for mpas e2e tests. Something which you can not impose on external contributors.

If an open source contributor works on a test that involves multiple projects and one core element is updated then that's the thing that you have to do in order to test those versions.

>Instead, an automation will do this process and notify of a failure due to the changes.

Yes, this is what the e2e test suite is there for.

>Why do we need an automation and why cant we rely on a developer to detect the changes when they update the cli in controllers to the latest version? Because nobody wants landmines and wasting a day before starting a new ticket.

Because updating a dependency should never be done automatically before a human verifies everything is okay. Besides, dependabot exists. It is there to update dependencies. And the e2e suite is always testing latest versions together with each other. That is its purpose.

>If you look at how open source development works at Apache, the tests are run on the code before the Apache collaborator even reviews the code that is written by an external contributor for a ticket. We need this process.

Sure. That is the e2e testing framework.

**Comment by robertwol** [comment]
@shivenduverma-sap can we slice and dice this issue to some managable chunks?

**Comment by shivenduverma-sap** [comment]
1. All common dependencies in all projects should be on the same version, to have that we need a pipeline/Action to parse, compare and notify. It wont fail/block anything, it is informative
2. We need a CI solution to pick up latest changes and test with e2e tests to notify if there will be failures due to the possible changes. It wont block/fail the PR, it is also informative. Devs should not be required to run tests locally

**Comment by shivenduverma-sap** [comment]
https://github.com/open-component-model/ocm/issues/480
https://github.com/open-component-model/ocm/issues/481
